### PR TITLE
qa: reuse methods in vstart_runner.LocalCephManager

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: qa/tasks/ceph_manager.py [hunk 1]
```diff
@@ -1326,6 +1326,14 @@ def tmp(x):
             except CommandFailedError:
                 self.log('Failed to get pg_num from pool %s, ignoring' % pool)
 
+        self.testdir = teuthology.get_testdir(self.ctx)
+        self.run_cluster_cmd_prefix = [
+            'sudo', 'adjust-ulimits', 'ceph-coverage',
+            f'{testdir}/archive/coverage', 'timeout', '120', 'ceph',
```
Comment (tchaikov) [inline qa/tasks/ceph_manager.py:1332]:
```suggestion
            f'{self.testdir}/archive/coverage', 'timeout', '120', 'ceph',
```

Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py]: Thanks!  :)

**Comment by rishabh-d-dave** [comment]
The failure on make check CI job doesn't look related - https://jenkins.ceph.com/job/ceph-pull-requests/65333/. Retriggering the CI job.

**Comment by rishabh-d-dave** [comment]
jenkins test make check

File: qa/tasks/ceph_manager.py [hunk 2]
```diff
@@ -1329,7 +1329,7 @@ def tmp(x):
         self.testdir = teuthology.get_testdir(self.ctx)
         self.run_cluster_cmd_prefix = [
             'sudo', 'adjust-ulimits', 'ceph-coverage',
-            f'{testdir}/archive/coverage', 'timeout', '120', 'ceph',
+            f'{self.testdir}/archive/coverage', 'timeout', '120', 'ceph',
```
Comment (lxbsz) [inline qa/tasks/ceph_manager.py:1332]: @rishabh-d-dave 
Please move this fixing into the first commit. It's introduced by the first commit and then you're fixing here ?

Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1332]: Right. Fixed.

**Review by lxbsz — APPROVED** [review]
LGTM.

**Review by tchaikov — COMMENTED** [review]
the title of 

> qa/ceph_manager: make it possible to reuse few methods

does not sound right. as "few" practically means "no". probably you could revise it like

> qa/ceph_manager: define member vars in CephManager so they can be overridden
> 
> use the member variables in methods like run_ceph_w() and run_cluster_cmd()
> for building the commands, so they can be customized by the derived classes
> like `vstart_runner.LocalCephManager`. hence these methods can be reused.

File: qa/tasks/vstart_runner.py [hunk 3]
```diff
@@ -964,53 +966,18 @@ def __init__(self):
         # methods to work though.
         self.pools = {}
 
+        self.cephadm = False
+        self.testdir = None
+        self.run_cluster_cmd_prefix = self.run_ceph_w_prefix = \
+            [os.path.join(BIN_PREFIX, "ceph")]
```
Comment (tchaikov) [inline qa/tasks/vstart_runner.py:972]: might want to add a comment noting that these variable are overridden here so that parent class can pick them up.

Comment (batrick) [reply qa/tasks/vstart_runner.py:972]: @rishabh-d-dave ^

Comment (rishabh-d-dave) [reply qa/tasks/vstart_runner.py]: Add a note to both vstart_runner.py and ceph_manager.py

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
@batrick Should this one too have same QA coverage as previous ceph_manager PRs?

**Comment by batrick** [comment]
yes

**Comment by rishabh-d-dave** [comment]
@neha-ojha @tchaikov 

The changes on this PR passed FS integration tests -  
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-09_05:03:18-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-09_04:35:44-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/

Can you please include it in your upcoming integration test run?

**Comment by rishabh-d-dave** [comment]
@batrick Rebased this PR, resolved a conflict in vstart_runner.py.

File: qa/tasks/ceph_manager.py [hunk 4]
```diff
@@ -1363,7 +1363,6 @@ def __init__(self, controller, ctx=None, config=None, logger=None,
         self.controller = controller
         self.next_pool_id = 0
         self.cluster = cluster
-        self.cephadm = cephadm
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1366]: no need to move this

File: qa/tasks/ceph_manager.py [hunk 5]
```diff
@@ -1375,6 +1374,17 @@ def tmp(x):
             self.log = tmp
         if self.config is None:
             self.config = dict()
+
+        self.testdir = teuthology.get_testdir(self.ctx)
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1378]: Nothing uses this except the code below. Convert to local?

File: qa/tasks/ceph_manager.py [hunk 6]
```diff
@@ -1445,17 +1453,12 @@ def run_ceph_w(self, watch_channel=None):
...
-                '--cluster',
-                self.cluster,
-                "-w"]
+        args = self.run_ceph_w_prefix + ['-w']
         if watch_channel is not None:
             args.append("--watch-channel")
             args.append(watch_channel)
-        return self.controller.run(args=args, wait=False, stdout=StringIO(), stdin=run.PIPE)
+        return self.controller.run(args=args, wait=False, stdout=StringIO(),
+                                   stdin=run.PIPE)
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1461]: code churn, undo please

File: qa/tasks/vstart_runner.py [hunk 7]
```diff
@@ -777,57 +778,21 @@ def __init__(self):
         # methods to work though.
         self.pools = {}
 
+        self.testdir = None
```
Comment (batrick) [inline qa/tasks/vstart_runner.py:781]: ditto

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise, LGTM

File: qa/tasks/ceph_manager.py (hunk 4 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1366]: This variable too is being overriden; see https://github.com/ceph/ceph/pull/38481/files#diff-203644a2b9fd4a8cf54b0c52970ffbd1aebb87a134dcf9fc0dc88286da7d4ab6R1380. So, I moved it to the bunch where variables are being overriden in a subclass and added a comment; see https://github.com/ceph/ceph/pull/38481/files#diff-c0b3c977c8b56ffb318d67c2f39c109d3a097714adff0f7870217cabe79e2486R785.

File: qa/tasks/ceph_manager.py (hunk 5 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1378]: The class had few calls to `get_testdir()`, replaced them with this class variable.

File: qa/tasks/ceph_manager.py (hunk 6 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py]: Done.

File: qa/tasks/vstart_runner.py (hunk 7 above)
Comment (rishabh-d-dave) [reply qa/tasks/vstart_runner.py:781]: Marking resolved due to https://github.com/ceph/ceph/pull/38481#discussion_r592928680.

File: qa/tasks/ceph_manager.py [hunk 8]
```diff
@@ -1551,11 +1551,10 @@ def do_rados(self, remote, cmd, check_status=True):
         """
         Execute a remote rados command.
         """
-        testdir = teuthology.get_testdir(self.ctx)
         pre = [
             'adjust-ulimits',
             'ceph-coverage',
-            '{tdir}/archive/coverage'.format(tdir=testdir),
+            f'{self.testdir}/archive/coverage'
```
Comment (tchaikov) [inline qa/tasks/ceph_manager.py:1557]: might want to add a comma at the end of this line.  i have
```
2021-03-13T06:29:38.588 DEBUG:teuthology.orchestra.run.smithi152:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coveragetimeout 0 ceph --cluster ceph --admin-daemon /var/run/ceph/ceph-mon.a.asok config show
2021-03-13T06:29:38.613 INFO:teuthology.orchestra.run.smithi152.stderr:/bin/ceph-coverage: line 25: exec: 0: not found
2021-03-13T06:29:38.614 DEBUG:teuthology.orchestra.run:got remote process result: 127
```

when testing this change. see https://pulpito.ceph.com/kchai-2021-03-13_05:46:44-rados-wip-kefu-testing-2021-03-13-1213-distro-basic-smithi/5960455/

for more instances, see https://pulpito.ceph.com/kchai-2021-03-13_05:46:44-rados-wip-kefu-testing-2021-03-13-1213-distro-basic-smithi/

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: qa/tasks/ceph_manager.py (hunk 8 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py]: Fixed.

**Review by batrick — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
@batrick I tested again; the tests passed -
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-17_05:08:45-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-17_05:08:52-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-17_18:29:46-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/

BTW, you never posted QA results for this PR before/after removing the label wip-podnnell-testing. Maybe it got missed it?

**Comment by rishabh-d-dave** [comment]
> @batrick I tested again; the tests passed -
> http://pulpito.front.sepia.ceph.com/rishabh-2021-03-17_05:08:45-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/
> http://pulpito.front.sepia.ceph.com/rishabh-2021-03-17_05:08:52-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/
> http://pulpito.front.sepia.ceph.com/rishabh-2021-03-17_18:29:46-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/
> 
> Tested vstart_runner.py too by running a couple of cephfs tests locally.
> 
> BTW, you never posted QA results for this PR before/after removing the label wip-podnnell-testing. Maybe it got missed it?

@batrick ping

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
@batrick Tests were successful -
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-24_04:04:12-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-23_12:15:42-fs-wip-rishabh-vr-set-shell-to-True-distro-basic-smithi/

**Comment by rishabh-d-dave** [comment]
Rebased once again.

**Review by lxbsz — APPROVED** [review]

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-04-01_00:45:34-fs-wip-pdonnell-testing-20210331.222326-distro-basic-smithi/

- https://tracker.ceph.com/issues/48772
    qa: pjd: not ok 9, 44, 80
- https://tracker.ceph.com/issues/50177
    osd: "stalled aio... buggy kernel or bad device?"
- https://tracker.ceph.com/issues/48771
    qa: iogen: workload fails to cause balancing
- https://tracker.ceph.com/issues/49845
    qa: failed umount in test_volumes
- https://tracker.ceph.com/issues/48773
    qa: scrub does not complete
- https://tracker.ceph.com/issues/48805
    mds: "cluster [WRN] Scrub error on inode 0x1000000039d (/client.0/tmp/blogbench-1.0/src/blogtest_in) see mds.a log and `damage ls` output for details"
- https://tracker.ceph.com/issues/50178
    qa: "TypeError: run() got an unexpected keyword argument 'shell'"
- https://tracker.ceph.com/issues/45434
    qa: test_full_fsync (tasks.cephfs.test_full.TestClusterFull) failed

File: qa/tasks/ceph_manager.py [hunk 9]
```diff
@@ -1582,17 +1585,12 @@ def run_ceph_w(self, watch_channel=None, shell=False):
...
-                '--cluster',
-                self.cluster,
-                "-w"]
+        args = self.run_ceph_w_prefix + ['-w']
         if watch_channel is not None:
             args.append("--watch-channel")
             args.append(watch_channel)
-        return self.controller.run(args=args, wait=False, stdout=StringIO(), stdin=run.PIPE)
+        return self.controller.run(args=args, wait=False, stdout=StringIO(),
+                                   stdin=run.PIPE, shell=shell)
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1593]: https://tracker.ceph.com/issues/50178

Looks like this was the actual cause, not #38443 ?

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/tasks/ceph_manager.py (hunk 9 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1593]: > Looks like this was the actual cause, not #38443 ?

Yes. But AFAIS, teuthology's run.run() requires an update. Created a PR - https://github.com/ceph/teuthology/pull/1639.

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
The last changes on this PR was a very minor fix introduced against the bug exposed by this run of Ceph API test - https://jenkins.ceph.com/job/ceph-api/19386/. The fix was to add `self.rook = False` to `LocalCephManager.__init__()`

**Comment by rishabh-d-dave** [comment]
@batrick Please take a look.

File: qa/tasks/ceph_manager.py [hunk 10]
```diff
@@ -1609,17 +1613,12 @@ def run_ceph_w(self, watch_channel=None, shell=False):
...
         :type watch_channel: str
         """
-        args = ["sudo",
-                "daemon-helper",
-                "kill",
-                "ceph",
-                '--cluster',
-                self.cluster,
-                "-w"]
+        args = self.run_ceph_w_prefix + ['-w']
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1616]: You can also try:

`args = ['exec', 'sudo', ...]`

Which will cause the shell to `execve` the `ceph` command, effectively.

Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1617]: Tried this change and tested this PR with and without this change during last updates. `run_ceph_w()` still leads to a crash when `shell` is set to `False`, doesn't crash when `shell` is set to `True` and prepending `exec sudo` to command arguments also prevents the crash. I've merged this fix in respective commits on this PR.

If this PR looks okay to you, we can move to final QA testing IMO.

Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1617]: Links to last few runs of Ceph API CI job for this PR -

Tested the fix(adding `exec sudo` and setting shell to False) to command arguments; job passed - https://jenkins.ceph.com/job/ceph-api/19955/
Retested if `shell=True` also worked; job passed - https://jenkins.ceph.com/job/ceph-api/19968/
Retested that `shell=False` still leads to crash; job failed (expected) - https://jenkins.ceph.com/job/ceph-api/20024/
Tested the fix (removing `shell=True` and prepending cmd args with `exec sudo`)  merged in respective commit; job passed - https://jenkins.ceph.com/job/ceph-api/20089/

**Comment by rishabh-d-dave** [comment]
jenkins test api

File: qa/tasks/ceph_manager.py (hunk 10 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1615]: Not sure why Ceph API job was re-triggered but it failed - https://jenkins.ceph.com/job/ceph-api/20110/

File: qa/tasks/ceph_manager.py [hunk 11]
```diff
@@ -1515,8 +1515,7 @@ def __init__(self, controller, ctx=None, config=None, logger=None,
         self.controller = controller
         self.next_pool_id = 0
         self.cluster = cluster
-        self.cephadm = cephadm
-        self.rook = rook
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1519]: undo the movement of these lines

File: qa/tasks/ceph_manager.py [hunk 12]
```diff
@@ -1528,6 +1527,18 @@ def tmp(x):
             self.log = tmp
         if self.config is None:
             self.config = dict()
+
+        self.testdir = teuthology.get_testdir(self.ctx)
```
Comment (batrick) [inline qa/tasks/ceph_manager.py:1531]: this line change should be part of `qa/ceph_manager: minor refactor`

**Review by batrick — CHANGES_REQUESTED** [review]
and move `qa/ceph_manager: minor refactor` to separate PR.

File: qa/tasks/ceph_manager.py (hunk 11 above)
Comment (lxbsz) [reply qa/tasks/ceph_manager.py:1519]: I think @rishabh-d-dave just wants to move all the variables needed to be overridden together.

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
The current state of Ceph API tests CI job on this PR and test PRs (https://github.com/ceph/ceph/pull/42306 and https://github.com/ceph/ceph/pull/42307) is exactly as expected -

this PR   passed https://jenkins.ceph.com/job/ceph-api/22450/
#42306  passed https://jenkins.ceph.com/job/ceph-api/22467/
#42307  failed https://jenkins.ceph.com/job/ceph-api/22466/

Therefore, the bug that I was facing in last version of PR branch is definitely fixed.

File: qa/tasks/ceph_manager.py (hunk 11 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1519]: Exactly.

File: qa/tasks/ceph_manager.py (hunk 12 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py]: Fixed, thanks!

File: qa/tasks/ceph_manager.py (hunk 10 above)
Comment (rishabh-d-dave) [reply qa/tasks/ceph_manager.py:1616]: Implemented this solution, it works fine now.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-August-27

**Pull request merged**